File: python/examples/disassemble.py
#!/usr/bin/env python3
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# disassemble: List functions, pick one, and show its disassembly.
#
# Usage: python disassemble.py [host_url]
#
# Defaults: http://127.0.0.1:18080, expects a program already open in Ghidra.

import sys

import libghidra as ghidra


def main() -> None:
    url = sys.argv[1] if len(sys.argv) >= 2 else "http://127.0.0.1:18080"
    client = ghidra.connect(url)

    # 1. List functions to find a target
    try:
        func_resp = client.list_functions(limit=20)
    except ghidra.GhidraError as e:
        print(f"list_functions failed: {e}", file=sys.stderr)
        sys.exit(1)

    functions = func_resp.functions
    if not functions:
        print("No functions found -- is a program open?", file=sys.stderr)
        sys.exit(1)

    print(f"Functions ({len(functions)} shown):")
    for f in functions:
        params = f"({f.parameter_count} params)" if f.parameter_count else ""
        thunk = " [thunk]" if f.is_thunk else ""
        print(f"  0x{f.entry_address:08x}  {f.name:<32}  {f.size:>6} bytes  {params}{thunk}")

    # 2. Pick the largest non-thunk function for interesting disassembly
    candidates = [f for f in functions if not f.is_thunk and f.size > 0]
    target = max(candidates, key=lambda f: f.size) if candidates else functions[0]

    print(f"\nTarget: {target.name} at 0x{target.entry_address:08x} ({target.size} bytes)")
    if target.prototype:
        print(f"Prototype: {target.prototype}")

    # 3. Get the single instruction at the entry point
    try:
        single = client.get_instruction(target.entry_address)
    except ghidra.GhidraError as e:
        print(f"get_instruction failed: {e}", file=sys.stderr)
        sys.exit(1)

    if single.instruction:
        insn = single.instruction
        print(f"\nEntry instruction:")
        print(f"  0x{insn.address:08x}  {insn.disassembly}  ({insn.length} bytes)")

    # 4. List all instructions in the function range
    try:
        insn_resp = client.list_instructions(
            range_start=target.start_address,
            range_end=target.end_address,
        )
    except ghidra.GhidraError as e:
        print(f"list_instructions failed: {e}", file=sys.stderr)
        sys.exit(1)

    instructions = insn_resp.instructions
    print(f"\nDisassembly of {target.name} ({len(instructions)} instructions):")
    print(f"  {'Address':<18}  {'Mnemonic':<12}  {'Operands':<32}  {'Len':>3}")
    print(f"  {'-' * 18}  {'-' * 12}  {'-' * 32}  {'-' * 3}")

    for insn in instructions:
        print(f"  0x{insn.address:08x}        {insn.mnemonic:<12}  {insn.operand_text:<32}  {insn.length:>3}")

    # 5. Summary
    total_bytes = sum(i.length for i in instructions)
    unique_mnemonics = len(set(i.mnemonic for i in instructions))
    print(f"\nSummary: {len(instructions)} instructions, {total_bytes} bytes, {unique_mnemonics} unique mnemonics")

    # Mnemonic frequency
    freq: dict[str, int] = {}
    for insn in instructions:
        freq[insn.mnemonic] = freq.get(insn.mnemonic, 0) + 1
    top = sorted(freq.items(), key=lambda kv: kv[1], reverse=True)[:5]
    print("Top mnemonics:")
    for mnem, count in top:
        print(f"  {mnem:<12}  {count}")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: rust/examples/disassemble.rs
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// disassemble: List functions, pick one, and disassemble its instructions.
//
// Usage: disassemble [host_url]

use libghidra as ghidra;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let url = args.get(1).map_or("http://127.0.0.1:18080", |s| s.as_str());
    let client = ghidra::connect(url);

    // 1. List the first 20 functions
    let funcs = match client.list_functions(0, u64::MAX, 20, 0) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("ListFunctions failed (is a program open?): {e}");
            std::process::exit(1);
        }
    };

    if funcs.functions.is_empty() {
        eprintln!("No functions found in the current program.");
        std::process::exit(1);
    }

    println!("Functions ({} shown):", funcs.functions.len());
    for f in &funcs.functions {
        println!(
            "  0x{:08x}  {:<40} {} bytes, {} params",
            f.entry_address, f.name, f.size, f.parameter_count
        );
    }

    // 2. Pick the largest function for a more interesting disassembly
    let target = funcs.functions.iter().max_by_key(|f| f.size).unwrap();

    println!(
        "\nSelected '{}' (0x{:08x}, {} bytes) for disassembly.",
        target.name, target.entry_address, target.size
    );

    // 3. Get the single instruction at the entry point
    let entry_insn = client
        .get_instruction(target.entry_address)
        .unwrap_or_else(|e| {
            eprintln!("GetInstruction failed: {e}");
            std::process::exit(1);
        });

    if let Some(insn) = &entry_insn.instruction {
        println!(
            "\nEntry instruction: 0x{:08x}  {}  (len={})",
            insn.address, insn.disassembly, insn.length
        );
    } else {
        println!(
            "\nNo instruction at entry point 0x{:08x}.",
            target.entry_address
        );
    }

    // 4. List all instructions within the function's address range
    let listing = client
        .list_instructions(target.start_address, target.end_address, 500, 0)
        .unwrap_or_else(|e| {
            eprintln!("ListInstructions failed: {e}");
            std::process::exit(1);
        });

    println!(
        "\nDisassembly of '{}' ({} instructions):\n",
        target.name,
        listing.instructions.len()
    );

    for insn in &listing.instructions {
        println!("  0x{:08x}  {}", insn.address, insn.disassembly);
    }

    // 5. Also get the full function record to show prototype
    let func_detail = client
        .get_function(target.entry_address)
        .unwrap_or_else(|e| {
            eprintln!("GetFunction failed: {e}");
            std::process::exit(1);
        });

    if let Some(f) = &func_detail.function {
        println!("\nFunction details:");
        println!("  Name:       {}", f.name);
        println!("  Namespace:  {}", f.namespace_name);
        println!("  Prototype:  {}", f.prototype);
        println!("  Is thunk:   {}", f.is_thunk);
        println!("  Params:     {}", f.parameter_count);
        println!(
            "  Range:      0x{:08x}..0x{:08x} ({} bytes)",
            f.start_address, f.end_address, f.size
        );
    }

    println!("\nDone.");
}
